### PR TITLE
Update SYCL_INTEL_bf16_conversion.asciidoc

### PR DESCRIPTION
**Pull request by AlexeySotkin**
Align the document with changes made in https://github.com/intel/llvm/pull/4989

### CODE REVIEW

File: sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc [hunk 1]
```diff
@@ -132,15 +135,16 @@ public:
...
 
-  // Get bfloat16 as uint16.
-  operator storage_t() const;
-
-  // Convert to bool type
+  // Convert bfloat16 to bool type
   explicit operator bool();
 
+  // Get bfloat16 as raw bits.
+  storage_t raw() const;
```
Comment (gmlueck) [inline sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: The names of the functions `from_bits()` and `raw()` seem inconsistent.  The names should be more similar since they are doing corresponding conversions.  For example, they could be named `from_bits()` and `bits()` or `from_raw()` and `raw()`.

**Comment by JackAKirk** [comment]
Am I right in understanding that there is no reason this proposal can't be implemented for other backends (cuda)?
 The intel backend specific element of the implementation seems to be only via the spirv functions `__spirv_ConvertFToBF16INTEL` and ` __spirv_ConvertBF16ToFINTEL`.  Is this correct?

Would it be acceptable for us to add NVPTX backend implementations of these spirv functions?

For example __spirv_ConvertFToBF16INTEL (perhaps suitably renamed?) would call the cuda builtin that performs the conversion using round to nearest even (as per the proposal).

Thanks

**Comment by JackAKirk** [comment]
> Am I right in understanding that there is no reason this proposal can't be implemented for other backends (cuda)? The intel backend specific element of the implementation seems to be only via the spirv functions `__spirv_ConvertFToBF16INTEL` and ` __spirv_ConvertBF16ToFINTEL`. Is this correct?
> 
> Would it be acceptable for us to add NVPTX backend implementations of these spirv functions?
> 
> For example __spirv_ConvertFToBF16INTEL (perhaps suitably renamed?) would call the cuda builtin that performs the conversion using round to nearest even (as per the proposal).
> 
> Thanks

I think I see that the implementations of the spirv functions that are labelled INTEL are hidden.  I think I understand that they will remain intel specific: in this case, and assuming that this extension could be generalized, what would be the preferred way of generalizing this extension to support other vendor's backends?  A higher level spirv function that would call INTEL labelled spirv/other backend specific spirv depending on the backend compiled for?  

Thanks

File: sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc (hunk 1 above)
Comment (AlexeySotkin) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Looks reasonable to me. I'd prefer `from_bits()` and `bits()` variant. @GhGmc2 what do you think?

Comment (Pennycook) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Is there a reason to add these new functions instead of re-using `bit_cast`?  `bit_cast<uint16_t>(a)` looks to be equivalent to `a.raw()`, and `bit_cast<bfloat16>(a)` equivalent to `from_bits(a)`.

Comment (AlexeySotkin) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Hmm, I can't think of any specific reason and it seems like it really can work with `bit_cast` out of the box: http://cmplrexplorer.intel.com/z/KG4rfq
But it becomes a bit wordy, especially in case when we want to initialize `bfloat16` with a literal
`bfloat16 X = sycl::bit_cast<bfloat16>(0);`
doesn't work, we need to specify the `from` type
`bfloat16 Y = sycl::bit_cast<bfloat16, uint16_t>(0);`

@GhGmc2 any objections if we remove `raw` and `from_bits`?

**Comment by JackAKirk** [comment]
> 

I made/tested an example draft oneapi namespace extension supporting CUDA here: https://github.com/intel/llvm/pull/5393

File: sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc (hunk 1 above)
Comment (gujinghui) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: May I know any progress on this discussion?
Looks like it's blocked for about long while.

We are going to add bfloat16 support in PyTorch master code base, if JIRA CMPLRLLVM-36893 is fixed.
But looks like the progress of bfloat16 in DPC++ compiler is waiting for this PR.
Could you help on this? @AlexeySotkin  @Pennycook  @gmlueck 

For from_bits, looks like PyTorch BFloat16 class has same API.
For completeness of sycl::bfloat16, perhaps we can reserve it?
Or, we can make it as a wrapper of sycl::bit_cast<bfloat16, uint16_t>?

Comment (gmlueck) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: It seems like the PyTorch BFloat16 could very easily use `sycl::bit_cast`, so I don't think that one use case should drive the decision.  I think the main argument is whether `sycl::bit_cast<bfloat16, uint16_t>(0)` is too verbose.  I don't know how common it will be to create a `bfloat16` from a uint16 literal, so I don't have a strong opinion.  @Pennycook?

Comment (Pennycook) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: I don't feel strongly about this either, but if the concern is specific to creating from literals, a user-defined literal might be a good solution there.  Then users would have options depending on what they're trying to do:
```c++
// Convert x to bit-wise representation
uint16_t bits = sycl::bit_cast<uint16_t>(x);

// Convert bits back to a bfloat16
bfloat16 y = sycl::bit_cast<bfloat16>(bits);

// Create bfloat16 directly from a literal
bfloat16 z = 1.234_bf16;
```

Comment (gujinghui) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Then, our conclusion is we will have no from_bits in bfloat16 class.
For bit_cast, did we agree on user-defined solution?
@Pennycook  @gmlueck  @rdeodhar 

// Convert x to bit-wise representation
uint16_t bits = sycl::bit_cast<uint16_t>(x);

// Convert bits back to a bfloat16
bfloat16 y = sycl::bit_cast<bfloat16>(bits);

// Create bfloat16 directly from a literal
bfloat16 z = 1.234_bf16;

Comment (gujinghui) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Any conclusion?
@Pennycook @gmlueck @rdeodhar

Comment (rdeodhar) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: I think adding these two function is not strictly necessary so we should not add them.

Comment (gmlueck) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Makes sense.  If we decide later that we want them, it is easy to add things to the extension.

Comment (rdeodhar) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: Based on the earlier discussion we will not be proceeding with this change. So this PR can be closed, right?

Comment (gmlueck) [reply sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc:146]: No, I think there are still some parts of this PR that we want:

* Removes the  implicit conversion to `uint16_t`.
* Adds an implicit conversion to `sycl::half`.

**Comment by rdeodhar** [comment]
OK, I've updated the PR to remove from_bits() and raw() and kept the other changes.

**Comment by gmlueck** [comment]
> OK, I've updated the PR to remove from_bits() and raw() and kept the other changes.

It looks like `from_bits` and `raw` are still in the synopsis.

This PR also needs to be merged with the main branch and updated.  The extension spec is now at "sycl/doc/extensions/experimental/sycl_ext_oneapi_bfloat16.asciidoc".

**Comment by rdeodhar** [comment]
Moved the required changes to "sycl/doc/extensions/experimental/sycl_ext_oneapi_bfloat16.asciidoc".

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bfloat16.asciidoc [hunk 2]
```diff
@@ -135,13 +135,14 @@ public:
   bfloat16(const float &a);
   bfloat16 &operator=(const float &a);
 
-  // Convert from bfloat16 to float
+  // Convert bfloat16 to floating-point types
   operator float() const;
+  operator sycl::half() const;
 
   // Get bfloat16 as uint16.
   operator storage_t() const;
```
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_bfloat16.asciidoc:143]: Somehow this conversion function reappeared after your merge.

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bfloat16.asciidoc]: I think this was just a merge mistake, so I removed it in 37835ee9d094969b7d3516da3b8f2395b98b451c

**Comment by MrSidims** [comment]
Closing in favor of https://github.com/intel/llvm/pull/6524 .
